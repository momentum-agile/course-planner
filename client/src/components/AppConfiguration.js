import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { theme, ThemeProvider, CSSReset } from "@chakra-ui/core";
import { DndProvider } from "react-dnd";
import { HTML5Backend } from "react-dnd-html5-backend";

/**
 * Wrapper Component to wrap all the different configurations
 * the main App Component uses.
 */
const AppConfiguration = ({ children }) => {
    return (
        <DndProvider backend={HTML5Backend}>
            <ThemeProvider theme={customTheme}>
                <CSSReset />
                <Router>{children}</Router>
            </ThemeProvider>
        </DndProvider>
    );
};

/**
 * See: https://chakra-ui.com/getting-started
 */
const customTheme = {
    ...theme,
    fontSizes: {
        xs: "0.75rem",
        sm: "0.875rem",
        md: "1rem",
        // Use for Large input headers
        lg: "2.5rem",
        xl: "1.25rem",
        "2xl": "1.5rem",
        "3xl": "1.875rem",
        "4xl": "2.25rem",
        "5xl": "3rem",
        "6xl": "4rem",
    },
};

export default AppConfiguration;
