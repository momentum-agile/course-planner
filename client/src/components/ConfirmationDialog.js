import React from "react";
import {
    Button,
    AlertDialog,
    AlertDialogOverlay,
    AlertDialogContent,
    AlertDialogHeader,
    AlertDialogBody,
    AlertDialogFooter,
    useToast,
} from "@chakra-ui/core";

const ConfirmationDialog = ({ isOpen, onClose, confirm, itemType, item, detail, action, navigateTo }) => {
    const cancelRef = React.useRef();
    const toast = useToast();
    return (
        <>
            <AlertDialog isOpen={isOpen} leastDestructiveRef={cancelRef} onClose={onClose}>
                <AlertDialogOverlay />
                <AlertDialogContent>
                    <AlertDialogHeader fontSize="lg" fontWeight="bold">
                        {action} {itemType}
                    </AlertDialogHeader>

                    <AlertDialogBody>
                        Are you sure you want to {action && action.toLowerCase()} {detail || "this"}? You can't undo this action afterwards.
                    </AlertDialogBody>

                    <AlertDialogFooter>
                        <Button ref={cancelRef} onClick={onClose}>
                            Cancel
                        </Button>
                        <Button
                            variantColor="red"
                            onClick={() => {
                                switch (itemType) {
                                    case "Course":
                                        confirm(item.courseCode);
                                        break;
                                    case "Programme":
                                        confirm();
                                        break;
                                    case "Student":
                                        confirm(item.upi);
                                        break;
                                    case "Regulation":
                                        confirm(item);
                                        break;
                                    default:
                                        confirm(item);
                                        break;
                                }
                                toast({
                                    title: `${action} ${itemType}  `,
                                    description: `${itemType}${
                                        detail ? ` '${detail}' ` : " "
                                    }has been successfully ${action.toLowerCase()}d`,
                                    status: "success",
                                    duration: 5000,
                                    isClosable: true,
                                });
                                onClose();
                            }}
                            ml={3}
                        >
                            {action}
                        </Button>
                    </AlertDialogFooter>
                </AlertDialogContent>
            </AlertDialog>
        </>
    );
};

export default ConfirmationDialog;
