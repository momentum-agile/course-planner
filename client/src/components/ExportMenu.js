import React from "react";
import { Menu, MenuButton, MenuList, MenuGroup, MenuItem, Text, Flex, IconButton } from "@chakra-ui/core";
import { colors as c } from "../colors";

const ExportMenu = ({ onExportMarkdown, onExportJson }) => {
    return (
        <Flex>
            <Menu>
                {({ isOpen }) => (
                    <React.Fragment>
                        <MenuButton>
                            <IconButton
                                icon="download"
                                size="xs"
                                _active={{ bg: c.darkBlue, color: c.whiteGrey }}
                                _hover={isOpen ? { bg: c.darkBlue, color: c.whiteGrey } : { bg: c.lightGrey, color: c.black }}
                                bg={isOpen ? c.darkBlue : "none"}
                                color={isOpen ? c.whiteGrey : c.black}
                            />
                        </MenuButton>

                        <MenuList placement="bottom-end" minWidth="160px">
                            <MenuGroup>
                                {onExportMarkdown && (
                                    <MenuItem onClick={onExportMarkdown}>
                                        <Text>Export to Markdown</Text>
                                    </MenuItem>
                                )}
                                {onExportJson && (
                                    <MenuItem onClick={onExportJson}>
                                        <Text>Export to JSON</Text>
                                    </MenuItem>
                                )}
                            </MenuGroup>
                        </MenuList>
                    </React.Fragment>
                )}
            </Menu>
        </Flex>
    );
};

export default ExportMenu;
