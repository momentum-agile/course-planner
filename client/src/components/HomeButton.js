import React from "react";
import { Link } from "react-router-dom";
import { IconButton } from "@chakra-ui/core";
import { AiFillHome } from "react-icons/ai";

const HomeButton = () => {
    return (
        <Link to="/">
            <IconButton as={AiFillHome} variantColor="black" left="20px" top="20px" size="xs" _hover={{ transform: "scale(1.1, 1.1)" }} />
        </Link>
    );
};

export default HomeButton;
