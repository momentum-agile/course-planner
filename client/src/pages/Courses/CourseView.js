import React, { useEffect, useState } from "react";
import { Flex, Text } from "@chakra-ui/core";
import { FieldsPane, RequirementsPane } from "./editables";
import { colors as c } from "../../colors";

const fields = {
    code: "courseCode",
    name: "name",
    desc: "description",
    sem: "semester",
    pts: "points",
    placeholder: "isPlaceholder",
};

/**
 * Component for viewing/editing/creating a course
 *
 * @param {object} course The entire course object to display
 * @param {bool} isNew Whether the view should display a new course to create. Setting isNew to true also sets isEditing to true
 * @param {boolean} isEditing Whether the course is being edited. This will render the components as inputs rather than text
 * @param {boolean} onEdit Callback function to call when edit is clicked on the course
 * @param {funciton} onDelete Callback function to call when delete is clicked on the course
 * @param {funciton} cancelUpdateCourse Callback function to call when cancelling edits to a course
 * @param {function} updateCourse Callback function to call when a course is 'saved' after being edited
 * @param {function} prefillCourse Callback function to call when trying to auto-generate a course from the code
 */
const CourseView = ({ course, isNew, isEditing, onEdit, onDelete, cancelUpdateCourse, updateCourse, prefillCourse, setPrefill }) => {
    const [code, setCode] = useState("");
    const [name, setName] = useState("");
    const [desc, setDesc] = useState("");
    const [sem, setSem] = useState([]);
    const [pts, setPts] = useState(15);
    const [placeholder, setPlaceholder] = useState(false);

    const setField = {
        code: setCode,
        name: setName,
        desc: setDesc,
        sem: setSem,
        pts: setPts,
        placeholder: setPlaceholder,
    };

    useEffect(() => {
        const { points, semester, description, courseCode, name, isPlaceholder } = course;

        setCode(courseCode || "");
        setName(name || "");
        setDesc(description || "");
        setSem(semester || []);
        setPts(points || 15);
        setPlaceholder(isPlaceholder || false);
    }, [course]);

    const changeField = (field, value) => {
        setField[field](value);
    };

    const saveCourse = () => {
        const editedCourse = { ...course };
        editedCourse[fields.code] = code;
        editedCourse[fields.name] = name;
        editedCourse[fields.desc] = desc;
        editedCourse[fields.sem] = sem;
        editedCourse[fields.pts] = pts;
        editedCourse[fields.placeholder] = placeholder;

        updateCourse(editedCourse);
    };

    const setCoursePrefill = (prefill) => {
        setPrefill({ ...prefill, _id: course._id });
    };

    return (
        <Flex height="100vh" direction="column" pt="30px" pr="10px">
            <FieldsPane
                item={course}
                code={code}
                name={name}
                desc={desc}
                sem={sem}
                pts={pts}
                placeholder={placeholder}
                isNew={isNew}
                isEditing={isEditing}
                onChange={changeField}
                onEdit={onEdit}
                onDelete={onDelete}
                onCancel={cancelUpdateCourse}
                onSave={saveCourse}
                prefillCourse={prefillCourse}
                setPrefill={setCoursePrefill}
            />

            {isNew ? (
                <Flex justifyContent="center" marginTop="20%" color={c.whiteGrey}>
                    <Text fontStyle="italic">First Create the course to add requirements</Text>
                </Flex>
            ) : (
                <Flex mt="10px">
                    <RequirementsPane course={course} updateCourse={updateCourse} />
                </Flex>
            )}
        </Flex>
    );
};

export default CourseView;
