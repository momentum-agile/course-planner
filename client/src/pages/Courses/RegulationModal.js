import React, { useState, useEffect } from "react";
import {
    FormControl,
    Input,
    Modal,
    ModalOverlay,
    ModalContent,
    ModalHeader,
    ModalFooter,
    ModalBody,
    ModalCloseButton,
    Button,
    Stack,
    TagLabel,
    Tag,
    TagCloseButton,
    NumberInput,
    NumberInputField,
    NumberInputStepper,
    NumberIncrementStepper,
    NumberDecrementStepper,
    Flex,
    IconButton,
    Text,
    Switch,
    FormLabel,
    ButtonGroup,
} from "@chakra-ui/core";
import CreatableSelect from "react-select/creatable";
import ReactTooltip from "react-tooltip";
import { colors as c } from "../../colors";
import useCourses from "../Courses/useCourses";

const defaultReg = {
    type: "POINTS",
    points: 15,
    courses: [],
};

const customStyles = {
    option: (provided) => ({
        ...provided,
        backgroundColor: c.white,
        "&:hover": {
            backgroundColor: c.uoaBlue,
            color: c.white,
        },
        flex: 1,
    }),
    control: (provided) => ({
        width: "100%",
        backgroundColor: c.white,
        '[type="text"]': {
            color: c.black,
        },
        flex: 1,
        ...provided,
    }),
    container: (provided) => ({
        ...provided,
        flex: 1,
    }),
};

const RegulationModal = ({ isOpen, onClose, title, updateCourse, course, editReg }) => {
    const [currentReg, setCurrentReg] = useState(defaultReg);
    const [outputRegulation, setOutputRegulation] = useState([]);
    const [selectValue,setSelectValue] = useState([])

    // Text field
    const [noteValue, setNoteValue] = useState("");

    const addToolTip = "Please add an AND or OR statement before adding another course";

    const { data } = useCourses();

    useEffect(() => {
        if (editReg) {
            setOutputRegulation(editReg);
        }
    }, [editReg, isOpen]);

    // plus button
    const handleAddToRegArrayClick = (type) => {
        if (type === "POINTS") {
            setOutputRegulation([...outputRegulation, `${currentReg.points} points from ${currentReg.courses.join(", ")}`]);
            setSelectValue([])
        } else if (noteValue.length > 0 && type === "NOTE") {
            setOutputRegulation([...outputRegulation, currentReg.courses[0]]);
            setNoteValue("");
        }
        setCurrentReg(defaultReg);
    };

    // Removes a regulation from the regulations array
    const handleRegTagCloseClick = (reg) => {
        setOutputRegulation(outputRegulation.filter((i) => i !== reg));
    };

    // updates current regulation
    const updateReg = (field) => (value, type) => {
        const newReg = {
            ...currentReg,
            type: type,
            [field]: value,
        };
        setCurrentReg(newReg);
    };

    const handleNoteChange = (event) => {
        setNoteValue(event.target.value);
        updateReg("courses")([event.target.value], "NOTE");
    };

    const handleSelectChange = (newValue, actionMeta) => {
        if (actionMeta.action === "select-option" || actionMeta.action === "create-option") { 
            const latestCourses = newValue.map((x) => x.label);
            setSelectValue(newValue)
            handleCourseSelect(latestCourses);
        }

        if (actionMeta.action === "pop-value" || actionMeta.action === "remove-value" || actionMeta.action === "clear") {
            if (newValue === null) {
                updateReg("courses")([], "POINTS");
                setSelectValue(newValue)
            } else {
                const latestCourses = newValue.map((x) => x.label);
                setSelectValue(newValue)
                handleCourseSelect(latestCourses);
            }
        }
    };

    // // Adds course to the current regulation (down button)
    const handleCourseSelect = (latestCourses) => {
        updateReg("courses")(latestCourses, "POINTS");
    };

    /** Toggle input between adding courses and a note
     * true: courses
     * false: notes
     */
    const [toggleInput, setToggleInput] = useState(true);
    const handleToggleChange = () => {
        setToggleInput(!toggleInput);
    };

    const handleAnd = () => {
        setOutputRegulation([...outputRegulation, "AND"]);
    };

    const handleOr = () => {
        setOutputRegulation([...outputRegulation, "OR"]);
    };

    const handleConfirm = () => {
        switch (title) {
            case "Prerequisites":
                updateCourse({
                    ...course,
                    prerequisites: outputRegulation,
                });
                break;
            case "InformalEquivalents":
                updateCourse({
                    ...course,
                    informalEquivalents: outputRegulation,
                });
                break;
            case "Corequisites":
                updateCourse({
                    ...course,
                    corequisites: outputRegulation,
                });
                break;
            case "Restrictions":
                updateCourse({
                    ...course,
                    restrictions: outputRegulation,
                });
                break;
            default:
                break;
        }
        handleClose();
    };

    const handleClose = () => {
        onClose();
        setCurrentReg(defaultReg);
        setNoteValue("");
    };

    const isAddDisabled =
        currentReg.courses.length === 0
            ? true
            : outputRegulation.length > 0 &&
              outputRegulation[outputRegulation.length - 1] !== "OR" &&
              outputRegulation[outputRegulation.length - 1] !== "AND";

    return (
        <>
            <Modal id="regulation-modal" isOpen={isOpen} onClose={handleClose} size="full">
                <ModalOverlay />
                <ModalContent>
                    <ModalHeader>Add {title}</ModalHeader>
                    <ModalCloseButton />

                    <ModalBody>
                        {/* Course/Note toggle */}
                        <FormControl as="fieldset">
                            <Flex align="center">
                                <FormLabel m="2" textAlign="center" px="0">
                                    Note
                                </FormLabel>
                                <Switch id="course-note-switch" color="#CBD5E0" onChange={handleToggleChange} isChecked={toggleInput} />
                                <FormLabel m="2" textAlign="center" px="0">
                                    Courses
                                </FormLabel>
                            </Flex>

                            {toggleInput ? (
                                <>
                                    {/* Courses input */}
                                    <Flex m="2" flex="1">
                                        <NumberInput
                                            size="md"
                                            defaultValue={15}
                                            min={10}
                                            height="40px"
                                            width="13%"
                                            onChange={(value) => updateReg("points")(value, "POINTS")}
                                            value={currentReg.points}
                                            step={5}
                                        >
                                            <NumberInputField />
                                            <NumberInputStepper>
                                                <NumberIncrementStepper />
                                                <NumberDecrementStepper />
                                            </NumberInputStepper>
                                        </NumberInput>
                                        <Text mx="1">points from</Text>
                                        <CreatableSelect
                                            isMulti
                                            onChange={handleSelectChange}
                                            options={data.map((course) => ({ value: course, label: course.courseCode }))}
                                            placeholder="Courses"
                                            formatCreateLabel={(userInput) => `Add Ghost Course: ${userInput}`}
                                            styles={customStyles}
                                            autoSize={true}
                                            value={selectValue}
                                        />
                                        <IconButton
                                            aria-label="Add input"
                                            icon="add"
                                            isDisabled={isAddDisabled}
                                            data-tip
                                            data-for="addTooltip"
                                            onClick={() => handleAddToRegArrayClick("POINTS")}
                                        />
                                        <ReactTooltip id="addTooltip" place="top" effect="solid" disable={!isAddDisabled}>
                                            {addToolTip}
                                        </ReactTooltip>
                                    </Flex>
                                </>
                            ) : (
                                // Notes input
                                <Flex m="2">
                                    <Input
                                        id="text"
                                        placeholder="Note"
                                        value={noteValue}
                                        onChange={handleNoteChange}
                                        onKeyPress={(ev) => {
                                            const isD =
                                                !noteValue ||
                                                (outputRegulation.length > 0 &&
                                                    outputRegulation[outputRegulation.length - 1] !== "OR" &&
                                                    outputRegulation[outputRegulation.length - 1] !== "AND");
                                            if (ev.key === "Enter" && !isD) {
                                                handleAddToRegArrayClick("NOTE");
                                            }
                                        }}
                                    />
                                    <IconButton
                                        aria-label="Add input"
                                        icon="add"
                                        onClick={() => {
                                            handleAddToRegArrayClick("NOTE");
                                        }}
                                        isDisabled={
                                            !noteValue ||
                                            (outputRegulation.length > 0 &&
                                                outputRegulation[outputRegulation.length - 1] !== "OR" &&
                                                outputRegulation[outputRegulation.length - 1] !== "AND")
                                        }
                                    />
                                </Flex>
                            )}
                        </FormControl>

                        {/* AND/OR buttons */}
                        <Flex justify="flex-end">
                            <ButtonGroup m="2" spacing={4}>
                                <Button
                                    variantColor="teal"
                                    variant="solid"
                                    onClick={handleAnd}
                                    isDisabled={
                                        outputRegulation.length <= 0 ||
                                        outputRegulation[outputRegulation.length - 1] === "OR" ||
                                        outputRegulation[outputRegulation.length - 1] === "AND"
                                    }
                                >
                                    AND
                                </Button>
                                <Button
                                    variantColor="teal"
                                    variant="solid"
                                    onClick={handleOr}
                                    isDisabled={
                                        outputRegulation.length <= 0 ||
                                        outputRegulation[outputRegulation.length - 1] === "OR" ||
                                        outputRegulation[outputRegulation.length - 1] === "AND"
                                    }
                                >
                                    OR
                                </Button>
                            </ButtonGroup>
                        </Flex>

                        {/* Final list of regulations */}
                        <Stack
                            id="reg-stack"
                            spacing={4}
                            isInlineborderWidth="2px"
                            p={2}
                            h="60px"
                            m="2"
                            bg={c.whiteGrey}
                            overflowX="scroll"
                            flexWrap="wrap"
                        >
                            <Flex>
                                {outputRegulation.map((reg, i) => (
                                    <>
                                        <Tag size="sm" key={i} rounded="full" variant="solid">
                                            <TagLabel>{reg}</TagLabel>
                                            <TagCloseButton onClick={() => handleRegTagCloseClick(reg)} />
                                        </Tag>
                                    </>
                                ))}
                            </Flex>
                        </Stack>
                    </ModalBody>

                    <ModalFooter>
                        <Button variantColor="red" mr={3} onClick={handleClose}>
                            Close
                        </Button>
                        <Button variantColor="green" onClick={handleConfirm} isDisabled={outputRegulation.length <= 0}>
                            Confirm
                        </Button>
                    </ModalFooter>
                </ModalContent>
            </Modal>
        </>
    );
};

export default RegulationModal;
