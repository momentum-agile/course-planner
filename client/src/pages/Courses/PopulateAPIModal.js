import React, { useState, useEffect, useRef } from "react";
import {
    Button,
    AlertDialog,
    AlertDialogOverlay,
    AlertDialogContent,
    AlertDialogHeader,
    AlertDialogBody,
    AlertDialogFooter,
    Checkbox,
    CheckboxGroup,
    Flex,
    FormControl,
    Select,
    Spinner,
    Text,
    useToast,
} from "@chakra-ui/core";
import { SearchBar } from "../../components";
import { CoursePlannerClient } from "../../common";
import filter from "@mcabreradev/filter";
import _ from "lodash";

const PopulateAPIModal = ({ isOpen, onClose, confirm, navigateTo }) => {
    const cancelRef = useRef();
    const [subject, setSubject] = useState("");
    const [searchTerm, setSearchTerm] = useState("");
    const [data, setData] = useState([]);
    const [loading, setLoading] = useState(false);

    const toast = useToast();

    useEffect(() => {
        setLoading(true);

        if (subject.length === 0) {
            setLoading(false);
            return;
        }

        CoursePlannerClient.getUniCoursesForDegree(subject)
            .then((res) => {
                setData(_.sortBy(res, "courseCode"));
                setLoading(false);
            })
            .catch((err) => {
                if (_.isEmpty(err)) {
                    toast({
                        isClosable: true,
                        duration: 9000,
                        title: `Error`,
                        description: `University API is down ðŸ˜ž`,
                        status: "error",
                    });
                }

                setLoading(false);
            });
    }, [subject, toast]);

    const handleModalClose = () => {
        onClose();
        setSubject("");
        setSearchTerm("");
        setData([]);
    };

    const handleOnCheck = (key) => (courseCode) => {
        const newData = [...data];
        const courseIdx = newData.findIndex((course) => course.courseCode === courseCode);

        newData[courseIdx] = {
            ...data[courseIdx],
            [key]: !data[courseIdx][key],
        };

        // When course and overwrite are checked, when you uncheck the course, this will then also uncheck the overwrite
        if (key === "checked" && data[courseIdx]["overwrite"]) {
            newData[courseIdx] = {
                ...newData[courseIdx],
                overwrite: false,
            };
        }

        setData(newData);
    };

    const handleAllCourses = (selected) => {
        const newData = data.map((course) => ({
            ...course,
            checked: selected,
        }));

        setData(newData);
    };

    const handleOverWrite = (selected) => {
        const newData = data.map((course) => ({
            ...course,
            overwrite: course.inDatabase && course.checked ? selected : false,
        }));

        setData(newData);
    };

    return (
        <>
            <AlertDialog isOpen={isOpen} leastDestructiveRef={cancelRef} onClose={handleModalClose}>
                <AlertDialogOverlay />
                <AlertDialogContent>
                    <AlertDialogHeader fontSize="lg" fontWeight="bold">
                        Populate Courses
                    </AlertDialogHeader>

                    <AlertDialogBody>
                        <Text paddingBottom="10px" fontSize="xl">
                            Select degree to populate courses
                        </Text>
                        <FormControl>
                            <Select id="subject" placeholder="Select Subject" onChange={(e) => setSubject(e.target.value)}>
                                <option value="SOFTENG">SOFTENG</option>
                                <option value="COMPSCI">COMPSCI</option>
                                <option value="ENGGEN">ENGGEN</option>
                                <option value="ENGSCI">ENGSCI</option>
                                <option value="ELECTENG">ELECTENG</option>
                                <option value="COMPSYS">COMPSYS</option>
                                <option value="MECHENG">MECHENG</option>
                            </Select>
                        </FormControl>
                        {loading && (
                            <Flex direction="column" width="100%" justifyContent="center" alignContent="center">
                                <Flex width="100%" justifyContent="center" alignContent="center" marginTop="25px">
                                    <Spinner size="xl" />
                                </Flex>
                            </Flex>
                        )}
                        {!loading && data.length > 0 && (
                            <Flex direction="column">
                                <SearchBar
                                    value={searchTerm}
                                    searchCategory="Populate"
                                    onChange={(e) => setSearchTerm(e.target.value)}
                                    width="100%"
                                />
                                <Flex direction="row" p={2} width="100%">
                                    <Flex width="50%" direction="column" justify="center">
                                        <Text textAlign="center" mb={1} as="b">
                                            {" "}
                                            Courses{" "}
                                        </Text>
                                        <Flex width="100%" justify="center">
                                            <Button size="xs" onClick={() => handleAllCourses(true)}>
                                                Select All
                                            </Button>
                                            <Button size="xs" ml={2} onClick={() => handleAllCourses(false)}>
                                                Deselect All
                                            </Button>
                                        </Flex>
                                    </Flex>
                                    <Flex width="50%" direction="column" justify="center">
                                        <Text textAlign="center" mb={1} as="b">
                                            {" "}
                                            Overwrite?{" "}
                                        </Text>
                                        <Flex width="100%" justify="center">
                                            <Button size="xs" onClick={() => handleOverWrite(true)}>
                                                Select All
                                            </Button>
                                            <Button size="xs" ml={2} onClick={() => handleOverWrite(false)}>
                                                Deselect All
                                            </Button>
                                        </Flex>
                                    </Flex>
                                </Flex>
                                <CheckboxGroup variantColor="blue" maxHeight="500px" overflowY="scroll" ml={2} mt={2}>
                                    {filter(data, { courseCode: searchTerm }).map(({ courseCode, inDatabase, checked, overwrite }, idx) => (
                                        <Flex direction="row" justifyContent="around" ml={4} key={idx}>
                                            <Flex width="50%">
                                                <Checkbox
                                                    key={idx}
                                                    value={courseCode}
                                                    isChecked={checked}
                                                    onChange={(e) => handleOnCheck("checked")(courseCode)}
                                                >
                                                    {courseCode}
                                                </Checkbox>
                                            </Flex>
                                            <Flex ml={20}>
                                                {inDatabase && (
                                                    <Checkbox
                                                        key={`${idx}-${inDatabase}`}
                                                        value={courseCode}
                                                        isChecked={overwrite}
                                                        isDisabled={!checked}
                                                        onChange={(e) => handleOnCheck("overwrite")(courseCode)}
                                                    />
                                                )}
                                            </Flex>
                                        </Flex>
                                    ))}
                                </CheckboxGroup>
                            </Flex>
                        )}
                    </AlertDialogBody>

                    <AlertDialogFooter>
                        <Button ref={cancelRef} onClick={handleModalClose}>
                            Cancel
                        </Button>
                        <Button
                            isDisabled={subject === ""}
                            variantColor="blue"
                            onClick={() => {
                                confirm(subject, data, toast);
                                handleModalClose();
                            }}
                            ml={3}
                        >
                            Populate
                        </Button>
                    </AlertDialogFooter>
                </AlertDialogContent>
            </AlertDialog>
        </>
    );
};

export default PopulateAPIModal;
