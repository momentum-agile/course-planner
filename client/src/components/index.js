import AppConfiguration from "./AppConfiguration.js";
import OutlineButton from "./OutlineButton.js";
import ProgrammeCard from "./ProgrammeCard.js";
import SearchBar from "./SearchBar.js";
import ProgrammeRequirementsItem from "./ProgrammeRequirementsItem.js";
import Table from "./Table.js";
import InlineEdit from "./InlineEdit";
import ConfirmationDialog from "./ConfirmationDialog";
import SearchSelect from "./SearchSelect";
import SingleSelect from "./SingleSelect";
import MenuWrapper from "./MenuWrapper";
import OptionsMenu from "./OptionsMenu";
import ExportMenu from "./ExportMenu";
import SaveCancelButtonSet from "./SaveCancelButtonSet";

export {
    AppConfiguration,
    OutlineButton,
    ProgrammeCard,
    SearchBar,
    ProgrammeRequirementsItem,
    Table,
    InlineEdit,
    SearchSelect,
    SingleSelect,
    ConfirmationDialog,
    OptionsMenu,
    MenuWrapper,
    ExportMenu,
    SaveCancelButtonSet,
};
