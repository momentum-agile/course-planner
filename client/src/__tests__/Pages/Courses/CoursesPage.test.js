import React from "react";
import renderer from "react-test-renderer";
import { AppConfiguration } from "../../../components";
import { Courses } from "../../../pages";
import CourseView from "../../../pages/Courses/CourseView";
import EmptyCourse from "../../../pages/Courses/EmptyCourse";
import PopulateAPIModal from "../../../pages/Courses/PopulateAPIModal";
import RegulationModal from "../../../pages/Courses/RegulationModal";

test("<Courses />", () => {
    const tree = renderer
        .create(
            <AppConfiguration>
                <Courses />
            </AppConfiguration>,
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("UoA API Populate Courses Modal", () => {
    const tree = renderer
        .create(
            <AppConfiguration>
                <PopulateAPIModal />
            </AppConfiguration>,
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("Empty Courses", () => {
    const tree = renderer
        .create(
            <AppConfiguration>
                <EmptyCourse />
            </AppConfiguration>,
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("Modal to add Regulations for Courses", () => {
    const tree = renderer
        .create(
            <AppConfiguration>
                <RegulationModal />
            </AppConfiguration>,
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
