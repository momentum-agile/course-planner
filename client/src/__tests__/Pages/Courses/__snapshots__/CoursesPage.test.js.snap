// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Courses /> 1`] = `
<div
  className="css-bulmef"
>
  <div
    className="css-il7sii"
  >
    <div
      className="css-1u9fuhx"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          className="css-1hd2lmz"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          type="button"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
          />
        </svg>
      </a>
    </div>
    <div
      className="css-gmuwbf"
    >
      <h2
        className="css-1n3wmrq"
      >
        Courses
      </h2>
    </div>
    <div
      className="css-1kh8xrj"
    >
      <button
        className="css-y6x91b"
        onClick={[Function]}
        type="button"
      >
        + Create
      </button>
      <button
        className="css-whzvt6"
        onClick={[Function]}
        type="button"
      >
        + Populate From Uni API
      </button>
    </div>
    <div
      className="css-9jn3o8"
    >
      <div
        className="css-1brfyz3"
      >
        <div
          className="css-1lads1q"
        >
          <div
            className="css-aetr73"
          >
            <div
              className="css-k6882c"
            >
              <svg
                className="css-1xb1rov"
                focusable={false}
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <input
              className="css-n5gi9c"
              onChange={[Function]}
              placeholder="Search by Course Code, Name, Semester or Points"
              type="search"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-nc95w2"
    >
      <div
        className="table css-1bdzjts"
      >
        <table
          className="css-8atqhb"
          role="table"
        >
          <thead
            className="css-0"
          >
            <tr
              className="css-1f81fq5"
              role="row"
            >
              <th
                className="css-11dd6zr"
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <p
                  className="css-7woomv"
                >
                  Course Code 
                </p>
              </th>
              <th
                className="css-11dd6zr"
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <p
                  className="css-7woomv"
                >
                  Course Name 
                </p>
              </th>
              <th
                className="css-11dd6zr"
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <p
                  className="css-7woomv"
                >
                  Semester(s) 
                </p>
              </th>
              <th
                className="css-11dd6zr"
                colSpan={1}
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "cursor": "pointer",
                  }
                }
                title="Toggle SortBy"
              >
                <p
                  className="css-7woomv"
                >
                  Points 
                </p>
              </th>
            </tr>
          </thead>
          <tbody
            className="css-0"
            role="rowgroup"
          />
        </table>
        <div
          className="css-ol8uv0"
        >
          <p
            className="css-1mg9g8m"
          >
            No data found
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-mzeyh"
  >
    <div
      className="css-lt6aye"
    >
      <p
        className="css-1fwnx19"
      >
        Create or select a Course to view its information here
      </p>
    </div>
  </div>
</div>
`;

exports[`Empty Courses 1`] = `
<div
  className="css-lt6aye"
>
  <p
    className="css-1fwnx19"
  >
    Create or select a Course to view its information here
  </p>
</div>
`;

exports[`Modal to add Regulations for Courses 1`] = `null`;

exports[`UoA API Populate Courses Modal 1`] = `null`;
